src/gencost/waterfall.py

```diff
@@ -2134,125 +2209,269 @@ def validate_merge_all_results(merge_all_df):
-                "step": Column(int, Check.isin((1, 2, 3))),
-                "report_year": Column(int, nullable=True),
-                "fuel_category": Column(str),
-                "capacity_mw": Column(float, Check.in_range(1e-1, 1e4)),
-                "gross_cf": Column(float, Check.ge(0.0), nullable=True),
-                "generator_starts": Column(int, Check.ge(0)),
-                "cum_starts": Column(int, Check.ge(0)),
-                "pollution_control_costs_per_kw": Column(float, Check.ge(0.0)),
-                "real_pollution_control_costs_per_kw": Column(float, Check.ge(0.0)),
-                "wage_scale": Column(float),
-            }
-            | {
-                "age_of_observation_secular_adj": Column(float),
-                "age_of_observation": Column(float, Check.in_range(0.0, 2e3)),
-                "age_relative_to_prime_avg": Column(float),
-            }
-            | {f"{k}_fraction": Column(float, Check.in_range(0.0, 1.0)) for k in fuels}
-            | {"minor_fuels_fraction": Column(float, Check.in_range(0.0, 1.0))}
-            | {k: Column(float, Check.in_range(0.0, 1.0)) for k in techs}
-            | {
-                k: Column(float, Check.gt(0.0), nullable=True)
-                for k in ("capex", "real_opex")
-            }
-            # not used in regression
-            | {
-                "age_in_report_year": Column(float),
-                "age_in_current_year": Column(float, Check.in_range(0.0, 2e3)),
-                "parasitic_load_pct": Column(float),
-                "camd_capacity_mw": Column(float, Check.in_range(0.0, 1e4)),
-                "gross_generation_mwh": Column(float, Check.ge(0.0)),
-                "gross_hr": Column(float, Check.ge(0.0), nullable=True),
-                "heat_in_mmbtu": Column(float, Check.ge(0.0)),
-                "net_generation_mwh": Column(float),
-                "net_cf": Column(float, nullable=True),
-                "arc": Column(float, nullable=True),
-                "inflator_to_2021": Column(float),
-                "fuel_starts": Column(int, Check.ge(0)),
-                "opex": Column(float, Check.ge(0.0), nullable=True),
-                "real_capex": Column(float, Check.ge(0.0), nullable=True),
-                "opex_per_kw": Column(float, Check.ge(0.0), nullable=True),
-                "capex_per_kw": Column(float, Check.ge(0.0), nullable=True),
-            }
-            | {f"{k}_mmbtu": Column(float, Check.ge(0.0), nullable=True) for k in fuels}
-            | {f"{k}_net_mwh": Column(float, nullable=True) for k in fuels}
-            | {
-                f"{k}_gross_mwh": Column(float, Check.ge(0.0), nullable=True)
-                for k in fuels
-            }
-            | {f"{k}_gross_cf": Column(float, Check.ge(0.0)) for k in fuels}
-        )
 
-        def gross_ge_net(df_):
-            return df_.gross_generation_mwh >= df_.net_generation_mwh
+        if level == "generator":
```

Comment:
I initially set up the pandera schema this way thinking that in the "else" statement I could do a check on the "core columns" in the generator level schema and then sub-plant specific columns, but I got an error when I tried to do something like 

schema = pa.DataFrameSchema(
                 columns=core_columns,
                columns=merge_all_columns,..)


in the else statement.

so the current structure of this validation function was a scrappy version to try to create a schema that worked for both merge_all and get_exa_by_generator

Reply:
Right, so you have to combine the dictionaries of columns and pass them as a single argument `columns`. The easiest way to do that would be something like this:

```python
schema = pa.DataFrameSchema(
    columns=core_columns | merge_all_columns,
    ...
)
```